File: problemset/Count_and_Say.cpp
char buff[102400];
class Solution {
public:
    std::vector<std::string> ret;
    string countAndSay(int n) {
      ret.push_back("1");

      for (int k = 1; k < 30; ++k) {
        const auto& str = ret.back();
        std::string now;
        int m = str.length();
        int cnt = 0;
        for (int i = 0; i < m; ++i) {
          if (!i || str[i] == str[i-1]) ++cnt;
          else {
            now += std::to_string(cnt);
            now += std::string(1, str[i-1]);
            cnt = 1;
          }
        }
        now += std::to_string(cnt);
        now += std::string(1, str.back());
        ret.push_back(now);
      }

      return ret[n-1];
    }
};

File: problemset/Count_and_Say.py
class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        ret1 = "1"
        ret2 = []
        for turn in xrange(2, n+1):
            length, cnt = len(ret1), 0
            for i in xrange(length):
                if i == 0:
                    cnt = 1
                elif ret1[i] == ret1[i-1]:
                    cnt += 1
                else:
                    ret2.extend([str(cnt), str(ret1[i-1])])
                    cnt = 1
            ret2.extend([str(cnt), str(ret1[-1])])
            ret1 = "".join(ret2)
            ret2 = []

        return ret1



